jit/gadgets-x86_64/control.S

```diff
@@ -16,15 +20,44 @@
     write_prep 32, call_indir
     movl 8(%_ip), %r14d
     movl %r14d, (%_addrq)
+    shrw $4, %r14w
+    movzwl %r14w, %r14d
+    leaq 16(%_ip), %r15
+    movq %r15, LOCAL_ret_cache(%_cpu, %r14, 8)
     write_done 32, call_indir
     movl %_tmp, %_eip
     jmp jit_ret
 
+.gadget call_postamble
+    movl (%_ip), %r14d
+    movq 8(%_ip), %r15
+    cmpl %r14d, %tmpd
+    jnz 1f
+    movq 16(%_ip), %_ip
+    btq $63, %_ip
+    jc 1f
+    leaq -JIT_BLOCK_code(%_ip), %r15
+    movq %r15, LOCAL_last_block(%_cpu)
+    gret
+1:
+    movq %r15, LOCAL_last_block(%_cpu)
+    movl %tmpd, %_eip
+    jmp jit_ret
+
 .gadget ret
     movl %_esp, %_addr
     addl 8(%_ip), %_esp
     read_prep 32, ret
-    movl (%_addrq), %_eip
+    movl (%_addrq), %tmpd
+    movl %tmpd, %r14d
+    shrw $4, %r14w
+    movzwq %r14w, %r14
+    movq LOCAL_ret_cache(%_cpu, %r14, 8), %_ip
+    cmpq $0, %_ip
+    jz 1f
+    gret
```

Comment:
Since this line will always go to a call_postamble gadget, it would make more sense to just write a jmp. It would then make more sense for call_postamble to be a function instead of a gadget, and make the call_postamble arguments into call arguments.

Reply:
Or make call_postamble a part of ret, and avoid the jump entirely.

In more detail: rearrange the arguments to call and call_indir so saved_ip, fake_ip, and current block are the first three. Current block is ignored by the call gadgets. ret updates _ip like it does now, then does the job of call_postamble, reading call arguments relative to _ip.

Reply:
OK. Makes sense to me.